resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesExternalShuffleService.scala

```diff
@@ -0,0 +1,315 @@
+        logInfo(s"Map: " +
+          s"${globalPartitionLengths.get((appId, shuffleId, mapId)).toString()}" +
+          s"for partitionId: $partitionId")
+        try {
+          val shuffleIndexInformation = shuffleIndexCache.get(indexFile)
+          val shuffleIndexRecord = shuffleIndexInformation.getIndex(partitionId)
+          val managedBuffer = new FileSegmentManagedBuffer(
+            transportConf,
+            getFile(appId, shuffleId, mapId, "data"),
+            shuffleIndexRecord.getOffset,
+            shuffleIndexRecord.getLength)
+          callback.onSuccess(managedBuffer.nioByteBuffer())
+        } catch {
+          case e: ExecutionException => logError(s"Unable to write index file $indexFile", e)
+        } finally {
+          responseDelayContext.stop()
+        }
+      case _ => super.handleMessage(message, client, callback)
+    }
+  }
+
+  protected override def handleStream(
+    header: BlockTransferMessage,
+    client: TransportClient,
+    callback: RpcResponseCallback): StreamCallbackWithID = {
+    header match {
+      case UploadParam(
+          appId, shuffleId, mapId, partitionId, partitionLength) =>
+        val responseDelayContext = metricSet.writeBlockRequestLatencyMillis.time()
+        try {
+          logInfo(s"Received upload param from app $appId")
+          val lengthMap = TreeMap(partitionId -> partitionLength.toLong)
+          globalPartitionLengths.merge((appId, shuffleId, mapId), lengthMap,
+            new BiFunction[TreeMap[Int, Long], TreeMap[Int, Long], TreeMap[Int, Long]]() {
+              override def apply(t: TreeMap[Int, Long], u: TreeMap[Int, Long]):
+                  TreeMap[Int, Long] = {
+                t ++ u
+              }
+            })
+          getFileWriterStreamCallback(
+            appId, shuffleId, mapId, "data", FileWriterStreamCallback.FileType.DATA)
+        } finally {
+          responseDelayContext.stop()
+        }
+      case _ =>
+        super.handleStream(header, client, callback)
+    }
+  }
+
+  protected override def getAllMetrics: MetricSet = metricSet
+
+  private def getFileWriterStreamCallback(
+      appId: String,
+      shuffleId: Int,
+      mapId: Int,
+      extension: String,
+      fileType: FileWriterStreamCallback.FileType): StreamCallbackWithID = {
+    val file = getFile(appId, shuffleId, mapId, extension)
+    val streamCallback =
+      new FileWriterStreamCallback(appId, shuffleId, mapId, file, fileType)
+    streamCallback.open()
+    streamCallback
+  }
+
+  private def getFile(
+      appId: String,
+      shuffleId: Int,
+      mapId: Int,
+      extension: String): File = {
+    Paths.get(shuffleDir.getAbsolutePath, appId,
+      s"shuffle_${shuffleId}_${mapId}_0.$extension").toFile
+  }
+
+  /** An extractor object for matching BlockTransferMessages. */
+  private object RegisterDriverParam {
+    def unapply(r: RegisterDriver): Option[(String, AppState)] =
```

Comment:
spark doesn't really use this pattern, of having object's with just `unapply` methods for pattern matching.  Its a bit harder for folks that have just a java background to follow (and arguably this is a part of the code where we want lower-level expertise, and too many scala-isms would discourage that).